internal/scheduler/metrics/metrics.go

```diff
@@ -392,3 +450,57 @@ func (m *Metrics) updateCounterVecFromJob(vec *prometheus.CounterVec, labels []s
 
 	return nil
 }
+
+// updateResourceSecondsCounterVec is a helper method to increment vector counters by the number of seconds per resource a jobs has consumed in a given state.
+func (m *Metrics) updateResourceSecondsCounterVec(vec *prometheus.CounterVec, labels []string, job *jobdb.Job) error {
+	i := len(labels)
+
+	// Number of jobs.
+	labels = append(labels, jobsResourceLabel)
+	if c, err := vec.GetMetricWithLabelValues(labels...); err != nil {
+		return err
+	} else {
+		c.Add(1)
+	}
+	requests := job.GetResourceRequirements().Requests
+	for _, resourceName := range m.config.TrackedResourceNames {
+		labels[i] = string(resourceName)
+		if c, err := vec.GetMetricWithLabelValues(labels...); err != nil {
+			return err
+		} else {
+			stateDuration := stateDuration(job.LatestRun(), labels[1]).Seconds()
+			q := requests[resourceName]
+			v := float64(q.MilliValue()) / 1000
+			c.Add(v * stateDuration)
+		}
+	}
+	return nil
+}
+
+func stateDuration(run *jobdb.JobRun, state string) time.Duration {
```

Comment:
This code assumes a particular set of state transitions, e.g., that only running jobs are preempted, which is not true in general. Would it not be easier to return the difference between the largest and second largest timestamp? Since the largest timestamp will correspond to the most recent state transition and the second largest to the one before that.

Reply:
Sounds good to me. I thought of this logic that may doesn't require sorting/comparing timestamps:
```
func getPriorState(run *jobdb.JobRun, state string) string {
	switch state {
	case queued:
		return ""
	case leased:
		return queued
	case pending:
		return leased
	case succeeded:
		return running
	case running:
		return pending
	case failed, cancelled, preempted:
		if run.Running() {
			return running
		} else if run.Pending() {
			return pending
		} else if run.Leased() {
			return leased
		} else {
			return queued
		}
	}
	return ""
}
```
what do you think of this approach ? also does it sounds logically correct ?

Reply:
I think the run will have already been marked as failed at this point.